File: Python/echo_client.py
import socket

def main():
    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        client.connect(('127.0.0.1', 8888))
        print("Đã kết nối đến Echo Server")
        
        while True:
            message = input("Bạn: ")
            
            if message.strip().lower() == 'quit':
                break
            
            client.send((message + '\n').encode('utf-8'))
            
            # Nhận echo từ server
            data = client.recv(1024)
            print(f"Echo từ server: {data.decode('utf-8').strip()}")
            
    except KeyboardInterrupt:
        print("\n Đang ngắt kết nối...")
    except Exception as e:
        print(f"Lỗi: {e}")
    finally:
        print(" Kết nối đã đóng")
        client.close()

if __name__ == "__main__":
    main()

File: C++/echo_client.cpp
#include <winsock2.h>
#include <iostream>
#pragma comment(lib, "ws2_32.lib")

#define PORT 8080
#define BUF_SIZE 4096

int main() {
    WSADATA wsa;
    SOCKET client;
    sockaddr_in serverAddr;
    char buffer[BUF_SIZE] = "Hello Echo TCP";

    WSAStartup(MAKEWORD(2,2), &wsa);

    client = socket(AF_INET, SOCK_STREAM, 0);

    int flag = 1;
    setsockopt(client, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));

    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(PORT);
    serverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    connect(client, (sockaddr*)&serverAddr, sizeof(serverAddr));

    send(client, buffer, strlen(buffer), 0);
    recv(client, buffer, BUF_SIZE, 0);

    std::cout << "Echo from server: " << buffer << "\n";

    closesocket(client);
    WSACleanup();
    return 0;
}
